src/main/java/uk/nhs/digital/nhsconnect/lab/results/inbound/NhsAckProducerService.java

```diff
@@ -0,0 +1,179 @@
+                Success currentMessageResult = (Success) messageProcessingResult;
+                MessageHeader messageHeader = currentMessageResult.getMessage().getMessageHeader();
+                nhsAckMessageContent = NhsAckMessageContent.builder()
+                        .messageNumber(messageHeader.getSequenceNumber())
+                        .messageStatus("MA")
+                        .build();
+                acceptedMessages++;
+            } else {
+                Error currentMessageResult = (Error) messageProcessingResult;
+                MessageHeader messageHeader = currentMessageResult.getMessage().getMessageHeader();
+                nhsAckMessageContent = NhsAckMessageContent.builder()
+                        .messageNumber(messageHeader.getSequenceNumber())
+                        .messageStatus("MR")
+                        .messageError(true)
+                        .messageErrorDescription(currentMessageResult.getException().getMessage())
+                        .build();
+                rejectedMessages++;
+            }
+            nhsAckContent.addMessageProcessingResult(nhsAckMessageContent);
+        }
+
+        if (acceptedMessages == totalMessages) {
+            nhsAckContent.setInterchangeStatusCode(NhsAckStatus.ACCEPTED);
+        } else if (rejectedMessages == totalMessages) {
+            nhsAckContent.setInterchangeStatusCode(NhsAckStatus.ALL_MESSAGES_REJECTED);
+        } else {
+            nhsAckContent.setInterchangeStatusCode(NhsAckStatus.PARTIALLY_ACCEPTED);
+        }
+
+        nhsAckContent.setSegmentCount(COMMON_SEGMENT_COUNT + acceptedMessages + (2 * rejectedMessages));
+
+        setDateTimes(nhsAckContent);
+
+        String segments = fillTemplate(nhsAckTemplate, nhsAckContent);
+
+        return segments.replace("\n", "");
+    }
+
+    public String createNhsAckEdifact(WorkflowId workflowId, InterchangeParsingException exception) {
+        Mustache nhsAckTemplate = loadTemplate("nhsAck.mustache");
+
+        var nhsAckContentBuilder = NhsAckContent.builder();
+
+        nhsAckContentBuilder.interchangeSender(exception.getSender());
+        nhsAckContentBuilder.interchangeRecipient(exception.getRecipient());
+        nhsAckContentBuilder.interchangeControlReference(String.valueOf(exception.getInterchangeSequenceNumber()));
+
+        nhsAckContentBuilder.nhsAckControlReference(String.valueOf(sequenceService.generateInterchangeSequence(
+                exception.getRecipient(), exception.getSender())));
+
+        nhsAckContentBuilder.workflowId(workflowId.getWorkflowId());
+
+        nhsAckContentBuilder.interchangeStatusCode(NhsAckStatus.INTERCHANGE_REJECTED);
+
+        nhsAckContentBuilder.interchangeError(true);
+        nhsAckContentBuilder.interchangeErrorDescription(exception.getMessage());
+
+        nhsAckContentBuilder.segmentCount(COMMON_SEGMENT_COUNT + 1);
+
+        NhsAckContent nhsAckContent = nhsAckContentBuilder.build();
```

Comment:
This can all be chained

```suggestion
        final var nhsAckContent = NhsAckContent.builder()
            .interchangeSender(exception.getSender())
            .interchangeRecipient(exception.getRecipient())
            .interchangeControlReference(String.valueOf(exception.getInterchangeSequenceNumber()))
            .nhsAckControlReference(String.valueOf(sequenceService.generateInterchangeSequence(
                exception.getRecipient(), exception.getSender())))
            .workflowId(workflowId.getWorkflowId())
            .interchangeStatusCode(NhsAckStatus.INTERCHANGE_REJECTED)
            .interchangeError(true)
            .interchangeErrorDescription(exception.getMessage())
            .segmentCount(COMMON_SEGMENT_COUNT + 1)
            .build();
```